include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,709 @@
+
+    /**
+     * Destructor.
+     */
+    ~ReadWriteVector ();
+
+    /**
+     * Sets the global size of the vector to @p size. The stored elements have
+     * their index in [0,size).
+     */
+    void reinit (const size_type size,
+                 const bool      fast = false);
+
+    /**
+     * Uses the same IndexSet as the one of the input vector @p in_vector and
+     * allocates memory for this vector.
+     *
+     * If the flag @p fast is set to false, the memory will be initialized
+     * with zero, otherwise the memory will be untouched (and the user must
+     * make sure to fill it with reasonable data before using it).
+     */
+    template <typename Number2>
+    void reinit(const ReadWriteVector<Number2> &in_vector,
+                const bool             fast = false);
+
+    /**
+     * Initializes the vector. The indices are specified by @p local_range,
+     */
+    void reinit (const IndexSet &local_range);
+
+    /**
+     * Swap the contents of this vector and the other vector @p v. One could
+     * do this operation with a temporary variable and copying over the data
+     * elements, but this function is significantly more efficient since it
+     * only swaps the pointers to the data of the two vectors and therefore
+     * does not need to allocate temporary storage and move data around.
+     *
+     * This function is analog to the the @p swap function of all C++
+     * standard containers. Also, there is a global function
+     * <tt>swap(u,v)</tt> that simply calls <tt>u.swap(v)</tt>, again in
+     * analogy to standard functions.
+     *
+     * This function is virtual in order to allow for derived classes to
+     * handle memory separately.
+     */
+    void swap (ReadWriteVector<Number> &v);
+
+    /**
+     * Copies the data and the IndexSet of the input vector @p in_vector.
+     */
+    ReadWriteVector<Number> &
+    operator = (const ReadWriteVector<Number> &in_vector);
+
+    /**
+     * Copies the data and the IndexSet of the input vector @p in_vector.
+     */
+    template <typename Number2>
+    ReadWriteVector<Number> &
+    operator = (const ReadWriteVector<Number2> &in_vector);
+
+#ifdef DEAL_II_WITH_PETSC
+    /**
+     * Imports all the elements present in the vector's IndexSet from the input
+     * vector @p petsc_vec.
+     */
+    ReadWriteVector<Number> &
+    operator = (const PETScWrappers::MPI::Vector &petsc_vec);
+#endif
+
+#ifdef DEAL_II_WITH_TRILINOS
+    /**
+     * Imports all the elements present in the vector's IndexSet from the input
+     * vector @p trilinos_vec.
+     */
+    ReadWriteVector<Number> &
+    operator = (const TrilinosWrappers::MPI::Vector &trilinos_vec);
+#endif
+
+    /**
+     * Sets all elements of the vector to the scalar @p s.
+     */
+    ReadWriteVector<Number> &operator = (const Number s);
+
+    /**
+     * Returns the global size of the vector, equal to number of elements in
+     * the vector.
+     */
+    size_type size () const;
```

Comment:
The comment should be clarified to make clear that it's not the elements that we store here, but the size of the index space.